# 4.5 Task - Data Consistency Checks

### Contents List

#### Practice work:
#### 1. Import libraries
#### 2. Importing Data
#### 3. Describe
#### 4. Mixed-type data
#### 5. Missing values
#### 6. Duplicate values
#### 7. Exporting Data frame
#### Task:
#### Question 2
#### Question 3
#### Question 4
#### Question 5
#### Question 6
#### Question 7
#### Question 8
#### Question 9
#### Question 10
#### Revisions
#### product_id investigation
#### Max price investigation
#### Exporting dataframes

#### Practice work:

#### 1. Import libraries

In [2]:
# Import libraries

import pandas as pd
import numpy as np
import os

#### 2. Import data

In [3]:
# declare path as variable

path = r'/Users/dekstrom/Documents/Career Foundry/Instacart Basket Analysis 11-2023'

In [94]:
# import products.csv as dataframe

df_prods = pd.read_csv(os.path.join(path,'02 Data','Original Data','products.csv'), index_col = False)

In [4]:
# import orders_wrangled.csv as dataframe

df_ords = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','orders_wrangled_1.csv'), index_col = False)

#### 3. Describe

In [5]:
# show overview of columns statistics

df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

#### 4. Mixed-type data

In [68]:
# create test data frame to practice on

df_test = pd.DataFrame()

In [70]:
# create a mixed-type data columns

df_test['mix'] = ['a', 'b', 1, True]

In [73]:

# show new data frame
df_test.head()

mix
0     a
1     b
2     1
3  True

In [76]:
# check for mixed-type columns in a data frame

for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


/var/folders/8w/2xbh8c1d2qnd_6yc30_xs1cw0000gp/T/ipykernel_7497/1384148929.py:4: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)


In [78]:
# check for mixed-type columns in a data frame
# use new version of DataFrame.applymap function

for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [79]:
# change data type

df_test['mix'] = df_test['mix'].astype('str')

#### 5. Missing values

In [95]:
# check for missing values and return count for each column

df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [96]:
# create a subset of df_prods made up of missing values

df_nan = df_prods[df_prods['product_name'].isnull()==True]

In [97]:
# show df_nan

df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [98]:
# show number of rows & columns

df_prods.shape

(49693, 5)

In [99]:
# create new data frame that excludes oservations with missing product names

df_prods_clean = df_prods[df_prods['product_name'].isnull()==False]

In [100]:
# show number of rows & columns

df_prods_clean.shape

(49677, 5)

#### 6. Duplicate values

In [101]:
# check for full duplicates

df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [102]:
# show duplicates

df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [103]:
# show number of rows & columns

df_prods_clean.shape

(49677, 5)

In [104]:
# remove full duplicates

df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [105]:
# show number of rows & columns

df_prods_clean_no_dups.shape

(49672, 5)

#### 7. Exporting data frame

In [106]:
# export data frame to csv

df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

#### Task:

#### Question 2

Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further. Tip: Keep an eye on min and max values!

In [6]:
# outline data frame statistics to analyze for anomalies
df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

##### Question 2 Answers

- There are fewer value counts for **days_since_prior_order** then for the the other columns. This suggests there are missing values that will need to be dealth with
- The mean is significantly larger than the median for **order_number** and **days_since_prior_order**. Though this doesn't mean we must act necessarily, it does suggests the presence of upper outliers in these variables which we may want to investigate.

#### Question 3

Check for mixed-type data in your df_ords dataframe.

In [7]:
# check for mixed-type columns in a df_ords

for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)
  else:
    print(col + ': not mixed')

order_id: not mixed
user_id: not mixed
order_number: not mixed
orders_day_of_week: not mixed
order_hour_of_day: not mixed
days_since_prior_order: not mixed


##### Question 3 Answers

- There are no columns with mixed-type data in df_ords.

#### Question 4

If you find mixed-type data, fix it. The column in question should contain observations of a single data type.

##### Question 4 Answers

- There are no columns with mixed-type data in df_ords.

#### Question 5

Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [8]:
# check for missing values and return count for each column

df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

In [9]:
# check whether order_number = 1 for days_since_prior_order = NaN to verify hypothesis that missing values
# indicate a user's first order
    
df_ords_prior_isnull = df_ords[df_ords['days_since_prior_order'].isnull()==True]

In [10]:
# verify that order_number = 1 for all days_since_prior_order = NaN

df_ords_prior_isnull['order_number'].describe()

count    206209.0
mean          1.0
std           0.0
min           1.0
25%           1.0
50%           1.0
75%           1.0
max           1.0
Name: order_number, dtype: float64

##### Question 5 Answers

- **days_since_prior_order** has 206,209 missing values. Generally, missing values are caused by either data corruption or not being recorded in first place. In this case, it's due to values not being recorded because **days_since_prior_order** only is null when the order is the first order from a user. In other words, the missing values indicate when order_number = 1.

#### Question 6

Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

In [11]:
# create flag column for days_since_prior_order = Nan

df_ords['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [12]:
# show df_ords

df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  first_order  
0                        8                     NaN         True  
1                        7                    15.0        False  
2                       12                    21.0        False  
3                        7                    29.0        False  
4                       15                    28.0        False  
...                    ...                     ...          ...  
3421078                 18                    29.0        False  
3421079                 10                    30.0        False  
3421080                 12                    18.0        False  
3421081                 12                     7.0        False  
3421082                 14                    30.0        False  

[3421083 rows x 7 columns]

##### Question 6 Answers

- I chose to create a flag column because we know that the nulls are important, they indicate an user's **first_order**. I could have chosen to not add a flag column since order_number = 1 indicates the same thing, but I added it anyway for better readability.
- Imputing values would be incorrect and removing the observations would change the overall description of the data.

#### Question 7

Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [13]:
# check for full duplicates

df_ords_dups = df_ords[df_ords.duplicated()]

In [14]:
# show duplicates

df_ords_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

#### Question 8

Address the duplicates using an appropriate method. In a markdown cell, explain why you used your method of choice

##### Question 8 Answers

- There are no full duplicates in df_ords.

#### Question 9

Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [15]:
# export data frame to csv

df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked_1.csv'))

#### Question 10

Save your Jupyter notebook and submit it here for your tutor to review.

#### Revisions

Check df_prods for data inconsistencies.

In [4]:
# import products.csv as dataframe

df_prods_revisions = pd.read_csv(os.path.join(path,'02 Data','Prepared Data','products_checked.csv'), index_col = 0)

In [5]:
# reset index

df_prods_revisions.reset_index(drop=True, inplace=True)

In [6]:
# Show tail of data frame

df_prods_revisions.tail()

product_id                                product_name  aisle_id  \
49667       49684   Vodka, Triple Distilled, Twist of Vanilla       124   
49668       49685          En Croute Roast Hazelnut Cranberry        42   
49669       49686                            Artisan Baguette       112   
49670       49687  Smartblend Healthy Metabolism Dry Cat Food        41   
49671       49688                      Fresh Foaming Cleanser        73   

       department_id  prices  
49667              5     5.3  
49668              1     3.1  
49669              3     7.8  
49670              8     4.7  
49671             11    13.5

In [7]:
# Check for data inconsistencies

df_prods_revisions.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

- The product_id variable has a higher value count than the max value by 16. Since this is a product table, we would expect the largest product_id to equal the value count (1 id per product). However, this difference of 16 is due to the 16 observations we eliminated already due to missing values. In any case, I will check duplicates in the product_id column since these should be unique.

- Secondly, the prices variable has a max of 99999 which is certainly an outlier and possible a mistake. I will investigate this as well.

**product_id investigation**

In [8]:
# check for partial duplicates in product_id column

df_prods_revisions_prodid_dups = df_prods_revisions[df_prods_revisions['product_id'].duplicated()]

In [9]:
# show partial duplicates in product_id column

df_prods_revisions_prodid_dups

product_id                                  product_name  aisle_id  \
6785         6800            Sprouted Quinoa Flakes Baby Cereal        92   
26505       26520  Cheese Shredded Sharp Cheddar Reduced Fat 2%        21   

       department_id  prices  
6785              18    14.0  
26505             16     2.9

In [10]:
# create dataframe with all observations that have non-unique product_ids

df_prods_revisions_prodid_dups = df_prods_revisions[df_prods_revisions['product_id'].isin([6800, 26520])]

In [11]:
# show dataframe with all observations that have non-unique product_ids

df_prods_revisions_prodid_dups

product_id                                       product_name  \
6784         6800                          Revive Zero Vitamin Water   
6785         6800                 Sprouted Quinoa Flakes Baby Cereal   
26504       26520  Clinical Advanced Solid Ultimate Fresh Anti-Pe...   
26505       26520       Cheese Shredded Sharp Cheddar Reduced Fat 2%   

       aisle_id  department_id  prices  
6784         64              7     6.4  
6785         92             18    14.0  
26504        80             11    10.6  
26505        21             16     2.9

In [12]:
# show observations surrounding duplicate product_ids

df_prods_revisions[df_prods_revisions['product_id'].isin([6798, 6799, 6800, 6801, 26518, 26519, 26520, 26521])]

product_id                                       product_name  \
6783         6798          18-in-1 Hemp Eucalyptus Pure Castile Soap   
6784         6800                          Revive Zero Vitamin Water   
6785         6800                 Sprouted Quinoa Flakes Baby Cereal   
6786         6801    Family Pack Bites Juicy Red & Green Apple Mixed   
26503       26518                      Chocolate Coconut Protein Bar   
26504       26520  Clinical Advanced Solid Ultimate Fresh Anti-Pe...   
26505       26520       Cheese Shredded Sharp Cheddar Reduced Fat 2%   
26506       26521  Organic Grass-Fed Classic Mild Cheddar Macaron...   

       aisle_id  department_id  prices  
6783         25             11     9.0  
6784         64              7     6.4  
6785         92             18    14.0  
6786        123              4     9.8  
26503         3             19     1.8  
26504        80             11    10.6  
26505        21             16     2.9  
26506         4              9     9.0

##### product_id investigation recap

- There are 4 seperate items that share 2 product_ids. This seems to be a mistake because there doesn't seem to be any relation between the items that share a product_id. I correct the data frame so that it contains fully unique product_ids.

In [13]:
# copy dataframe

df_prods_revisions_prodid_cleaned = df_prods_revisions.copy()

In [14]:
# change first duplicate product_id

df_prods_revisions_prodid_cleaned.at[6784, 'product_id'] = 6799

In [15]:
# change first duplicate product_id

df_prods_revisions_prodid_cleaned.at[26504, 'product_id'] = 26519

In [16]:
# show changed product_ids

df_prods_revisions_prodid_cleaned[df_prods_revisions_prodid_cleaned['product_id'].isin([6798, 6799, 6800, 6801, 26518, 26519, 26520, 26521])]

product_id                                       product_name  \
6783         6798          18-in-1 Hemp Eucalyptus Pure Castile Soap   
6784         6799                          Revive Zero Vitamin Water   
6785         6800                 Sprouted Quinoa Flakes Baby Cereal   
6786         6801    Family Pack Bites Juicy Red & Green Apple Mixed   
26503       26518                      Chocolate Coconut Protein Bar   
26504       26519  Clinical Advanced Solid Ultimate Fresh Anti-Pe...   
26505       26520       Cheese Shredded Sharp Cheddar Reduced Fat 2%   
26506       26521  Organic Grass-Fed Classic Mild Cheddar Macaron...   

       aisle_id  department_id  prices  
6783         25             11     9.0  
6784         64              7     6.4  
6785         92             18    14.0  
6786        123              4     9.8  
26503         3             19     1.8  
26504        80             11    10.6  
26505        21             16     2.9  
26506         4              9     9.0

**Max price investigation**

In [27]:
df_prods_revisions_prodid_cleaned.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [28]:
# check largest item prices

df_prods_sorted = df_prods_revisions_prodid_cleaned.sort_values(by = 'prices', ascending = False)

In [29]:
# check output

df_prods_sorted.head()

product_id                      product_name  aisle_id  department_id  \
33649       33664             2 % Reduced Fat  Milk        84             16   
21538       21553  Lowfat 2% Milkfat Cottage Cheese       108             16   
19376       19391         Turkey Breast Tenderloins        49             12   
25564       25579     Naturally Smoked Trout Fillet        15             12   
40469       40486                   Chicken Tenders        49             12   

        prices  
33649  99999.0  
21538  14900.0  
19376     25.0  
25564     25.0  
40469     25.0

There are 2 large outliers. I will replace these with the average price of their respective food type.

In [52]:
# check similar items

df_prods_milk  = df_prods_sorted[df_prods_sorted['product_name'].str.endswith(' Milk') == True]

In [70]:
# check more specifically 2% milk

df_prods_2perc_milk = df_prods_milk[df_prods_milk['product_name'].str.contains('2%') == True]

In [71]:
# check output

df_prods_2perc_milk

product_id                                       product_name  \
38879       38895          Super Natural Organic Reduced Fat 2% Milk   
37904       37919                          Organic 2% Chocolate Milk   
17936       17951       2% Reduced Fat Ultra-Filtered Chocolate Milk   
32463       32478                                Reduced Fat 2% Milk   
19805       19820                      Grassmilk 2% Reduced Fat Milk   
41826       41843                        2% Organic Reduced Fat Milk   
49302       49319  Mozzarella String Cheese Made with 2% Reduced ...   
1722         1729                               2% Lactose Free Milk   
40158       40174                        2% Reduced Fat Organic Milk   
33557       33572                   Lactose Free 2% Reduced Fat Milk   
37780       37795                     Dairy Pure 2% Reduced Fat Milk   
39847       39863                                    Organic 2% Milk   
22595       22610                   Reduced Fat 2% Lactose-Free Milk   
23894       23909                                2% Reduced Fat Milk   
1932         1940                        Organic 2% Reduced Fat Milk   
33785       33800                        Organic 2% Plain Kefir Milk   
14867       14882                        2% Milkfat Reduced Fat Milk   
5770         5785                        Organic Reduced Fat 2% Milk   
15206       15221                                            2% Milk   
20836       20851                 Filtered Fresh 2% Reduced Fat Milk   
23947       23962                   Lactose Free Reduced Fat 2% Milk   
22198       22213                      2% Reduced Fat Chocolate Milk   
23067       23082                                  2% Chocolate Milk   
35899       35914                    Dha Omega 3 Reduced Fat 2% Milk   
31202       31217                         Low Fat 2% Evaporated Milk   
34170       34185                               Lactose Free 2% Milk   
19156       19171      100% Lactose Free Organic 2% Reduced Fat Milk   
28578       28593                      Reduced Fat 2% Chocolate Milk   
19569       19584             Dairy Pure 2% Milkfat Reduced Fat Milk   
45421       45438                         Select 2% Reduced Fat Milk   

       aisle_id  department_id  prices  
38879        84             16    14.7  
37904        84             16    14.6  
17936        84             16    14.2  
32463        84             16    13.9  
19805        84             16    12.6  
41826       100             21    12.5  
49302        21             16    11.8  
1722         84             16    11.6  
40158        84             16    11.4  
33557        91             16    10.5  
37780        84             16    10.1  
39847        84             16     9.4  
22595        91             16     9.4  
23894        84             16     9.2  
1932         84             16     9.1  
33785       120             16     9.0  
14867        84             16     8.4  
5770         84             16     7.9  
15206        84             16     7.0  
20836        84             16     5.8  
23947        84             16     5.1  
22198        84             16     4.6  
23067        84             16     3.5  
35899        84             16     3.3  
31202        17             13     2.7  
34170        91             16     2.3  
19156        84             16     1.9  
28578        84             16     1.7  
19569        84             16     1.7  
45421        84             16     1.2

In [72]:
# check for other cottage cheese items

df_prods_cotcheese = df_prods_sorted[df_prods_sorted['product_name'].str.contains('Cottage Cheese') == True]

In [73]:
# check output

df_prods_cotcheese

product_id                               product_name  aisle_id  \
21538       21553           Lowfat 2% Milkfat Cottage Cheese       108   
11579       11594      Organic Cottage Cheese Kalamata Olive       108   
38542       38558             Organic Low Fat Cottage Cheese       108   
37827       37842         Sea Salt Small Curd Cottage Cheese        21   
47088       47105           Organic Pineapple Cottage Cheese       100   
...           ...                                        ...       ...   
46193       46210  4% Milkfat Min. Small Curd Cottage Cheese       108   
41750       41767                       Cottage Cheese 22 Oz       108   
19868       19883               Small Curd Cottage Cheese 4%       108   
19444       19459                  Small Curd Cottage Cheese       108   
39415       39431          Cottage Cheese 4% Milkfat Minimum       108   

       department_id   prices  
21538             16  14900.0  
11579             16     15.0  
38542             16     13.7  
37827             16     13.5  
47088             21     12.6  
...              ...      ...  
46193             16      1.8  
41750             16      1.6  
19868             16      1.3  
19444             16      1.3  
39415             16      1.1  

[71 rows x 5 columns]

In [87]:
# define variable for avg milk price

milk_avg_price = df_prods_2perc_milk['prices'].mean()

In [89]:
# check output

milk_avg_price

8.036666666666665

In [91]:
# define variable for avg cottage cheese price

cotcheese_avg_price = df_prods_cotcheese['prices'][1:].mean()

In [92]:
# check output

cotcheese_avg_price

6.757142857142858

In [102]:
# impute milk outlier with milk_avg_price

df_prods_revisions_prodid_cleaned.at[33649, 'prices'] = milk_avg_price

In [104]:
# impute cheese outlier with cotcheese_avg_price

df_prods_revisions_prodid_cleaned.at[21538, 'prices'] = cotcheese_avg_price

In [107]:
# confirm changes

df_prods_revisions_prodid_cleaned.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349734     67.762442      11.728942      7.680425
std    14340.705310     38.315784       5.850779      4.199299
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000     25.000000

In [106]:
# confirm changes

df_prods_revisions_prodid_cleaned.sort_values(by = 'prices', ascending = False)

product_id                      product_name  aisle_id  department_id  \
19376       19391         Turkey Breast Tenderloins        49             12   
25564       25579     Naturally Smoked Trout Fillet        15             12   
21452       21467            Wild Caught Raw Shrimp        15             12   
9005         9020  Boneless Skinless Chicken Thighs        35             12   
40469       40486                   Chicken Tenders        49             12   
...           ...                               ...       ...            ...   
11292       11307                  Goat Cheese Logs         2             16   
32788       32803                 Regular Deodorant        80             11   
46606       46623                   Pie Pans, Large        10             17   
37022       37037                       Redseedless        24              4   
21448       21463      Vanilla Bean Light Ice Cream        37              1   

       prices  
19376    25.0  
25564    25.0  
21452    25.0  
9005     25.0  
40469    25.0  
...       ...  
11292     1.0  
32788     1.0  
46606     1.0  
37022     1.0  
21448     1.0  

[49672 rows x 5 columns]

##### Max price investigation recap

- I imputed two outliers with the average price for their respective food type.

#### Exporting dataframes

In [108]:
# export df_prods_revisions_prodid_cleaned to csv

df_prods_revisions_prodid_cleaned.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked_final.csv'))